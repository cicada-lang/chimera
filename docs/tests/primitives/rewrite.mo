rule conjunctiveNormalForm {
  not(not(x)) => x
  not(and(x, y)) => or(not(x), not(y))
  not(or(x, y)) => and(not(x), not(y))
  or(and(x, y), z) => and(or(x, z), or(y, z))
  or(x, and(y, z)) => and(or(x, y), or(x, z))
}

print conjunctiveNormalForm(quote not(and(not(A), not(B))))

print rewriteManySteps(3, conjunctiveNormalForm, quote not(and(not(A), not(B))))
print rewriteManySteps(10, conjunctiveNormalForm, quote not(and(not(A), not(B))))
