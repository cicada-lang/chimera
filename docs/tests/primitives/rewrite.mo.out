or(A, B)
[or(not(not(A)), not(not(B))), or(A, not(not(B))), or(A, B)]
[or(not(not(A)), not(not(B))), or(A, not(not(B))), or(A, B)]
